packages/core/src/impl/ObservableQuery.ts

```diff
@@ -0,0 +1,79 @@
+import {observable, Observable, Observer, Subscription} from '../observable'
+import {CommunicationObserver} from '../transport'
+import Cancellable from '../util/Cancellable'
+
+export type QueryExecutor = (consumer: CommunicationObserver<string>) => void
+
+type Decorator<T> = (observer: Observer<T>) => Observer<string>
+
+class QuerySubscription implements Subscription {
+  private cancellable?: Cancellable
+  private isClosed = false
+
+  public constructor(observer: Observer<string>, executor: QueryExecutor) {
+    try {
+      executor({
+        next: value => {
+          observer.next(value)
+        },
+        error: e => {
+          this.isClosed = true
+          observer.error(e)
+        },
+        complete: () => {
+          this.isClosed = true
+          observer.complete()
+        },
+        useCancellable: c => {
+          this.cancellable = c
+        },
+      })
+    } catch (e) {
+      this.isClosed = true
+      observer.error(e)
+    }
+  }
+
+  public get closed(): boolean {
+    return this.isClosed
+  }
+
+  public unsubscribe(): void {
+    this.cancellable?.cancel()
+  }
+}
+
+function noop(): void {}
+
+function completeObserver<T>(observer: Partial<Observer<T>>): Observer<T> {
+  const {next, error, complete} = observer
+
+  return {
+    next: next ? next.bind(observer) : noop,
+    error: error ? error.bind(observer) : noop,
+    complete: complete ? complete.bind(observer) : noop,
+  }
+}
+
+export default class ObservableQuery<T> implements Observable<T> {
+  public constructor(
+    private readonly executor: QueryExecutor,
+    private readonly decorator: Decorator<T>
+  ) {}
+
+  public subscribe(
+    observerOrNext?: Partial<Observer<T>> | ((value: T) => void)
```

Comment:
what about error and complete optional parameters?

Reply:
resolving here, see the discussion on the observable typings